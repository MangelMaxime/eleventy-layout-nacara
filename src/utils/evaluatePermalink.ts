
export const evaluatePermalink = (ctx: any) => {
    let permalink = null;
    if (ctx.data.permalink) {
        if (typeof ctx.data.permalink === "function") {
            permalink = ctx.data.permalink(ctx.data)
        } else {
            permalink = ctx.data.permalink;
        }
    }
    return permalink;
}
