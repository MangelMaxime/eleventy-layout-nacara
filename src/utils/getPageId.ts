import path from "path";
// import NacaraTypes from "./../../types/nacara";

export default function getPageId(fileStem : string) {
    //  Normal the path, so we can split using the path separator
    const normalizedInputPath = path.normalize(fileStem);
    // Extract all the segments of the path
    const inputPathSegments = normalizedInputPath.split(path.sep);
    // console.log("inputPathSegments:", inputPathSegments);
    // Build the section direction, which consist of the root + the first segment of the path
    const pageIdSegments = inputPathSegments.slice(2);

    return pageIdSegments.join("/");
};
