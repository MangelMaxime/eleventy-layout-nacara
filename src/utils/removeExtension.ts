import path from "path";

export function removeExtension(fileName: string): string {
    const parsedPath = path.parse(fileName);
    return path.join(parsedPath.dir, parsedPath.name);
}
