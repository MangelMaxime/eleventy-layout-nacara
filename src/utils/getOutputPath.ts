import { evaluatePermalink } from "./evaluatePermalink.js";

export default function getOutputPath(pageContext: any) {
    const trimOutputRegex = new RegExp(`^${pageContext.data.eleventy.directories.output}`);
    const permalink = evaluatePermalink(pageContext);

    if (permalink) {
        return permalink;
    }

    return pageContext.data.page.outputPath.replace(trimOutputRegex, "");
}
