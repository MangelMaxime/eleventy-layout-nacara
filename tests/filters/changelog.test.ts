import { expect, test } from 'vitest'
const Eleventy = require("@11ty/eleventy");
import { formatHTML } from "../utils/_formatHTML";

test("known categories are rendered in a fixed order", async () => {
    const elev = new Eleventy(
        "./fixtures/changelog-0/",
        "./fixtures/changelog-0/_site",
        {
            configPath: "./fixtures/changelog-0/.eleventy.js",
        }
    );

    const json = await elev.toJSON();
    const result =
        json.find(
            (item: any) => item.url === "/docs/changelog/"
        );
    const formattedResult = formatHTML(result.content);

    expect(formattedResult).toMatchSnapshot();
});

test("unknown categories are rendered in alphabetical order", async () => {
    const elev = new Eleventy(
        "./fixtures/changelog-1/",
        "./fixtures/changelog-1/_site",
        {
            configPath: "./fixtures/changelog-1/.eleventy.js",
        }
    );

    const json = await elev.toJSON();
    const result =
        json.find(
            (item: any) => item.url === "/docs/changelog/"
        );
    const formattedResult = formatHTML(result.content);

    expect(formattedResult).toMatchSnapshot();
});

test("unknown categories are rendered after known categories", async () => {
    const elev = new Eleventy(
        "./fixtures/changelog-2/",
        "./fixtures/changelog-2/_site",
        {
            configPath: "./fixtures/changelog-2/.eleventy.js",
        }
    );

    const json = await elev.toJSON();
    const result =
        json.find(
            (item: any) => item.url === "/docs/changelog/"
        );
    const formattedResult = formatHTML(result.content);

    expect(formattedResult).toMatchSnapshot();
});

test("all the versions are rendered", async () => {
    const elev = new Eleventy(
        "./fixtures/changelog-3/",
        "./fixtures/changelog-3/_site",
        {
            configPath: "./fixtures/changelog-3/.eleventy.js",
        }
    );

    const json = await elev.toJSON();
    const result =
        json.find(
            (item: any) => item.url === "/docs/changelog/"
        );
    const formattedResult = formatHTML(result.content);

    expect(formattedResult).toMatchSnapshot();
});

test("body of a category is rendered convert from markdown to HTML", async () => {
    const elev = new Eleventy(
        "./fixtures/changelog-4/",
        "./fixtures/changelog-4/_site",
        {
            configPath: "./fixtures/changelog-4/.eleventy.js",
        }
    );

    const json = await elev.toJSON();
    const result =
        json.find(
            (item: any) => item.url === "/docs/changelog/"
        );
    const formattedResult = formatHTML(result.content);

    expect(formattedResult).toMatchSnapshot();
});
