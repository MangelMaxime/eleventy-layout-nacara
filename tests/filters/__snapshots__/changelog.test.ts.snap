// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`additional Markdown plugins are supported when rendering the markdown content 1`] = `
"<ul class="changelog-list">
    <li class="changelog-list-item is-version">
        <a href="#1.0.0-2021-07-27">
            <span class="anchor" id="1.0.0-2021-07-27"></span>
            <span class="tag is-primary is-large has-text-weight-bold">
                1.0.0
            </span>
        </a>
        <span class="release-date is-uppercase has-text-weight-bold is-size-5">
            27 July 2021
        </span>
    </li>
    <div>
        <li class="changelog-list-item">
            <span
                class="tag changelog-list-item-category is-medium has-text-weight-bold is-success"
            >
                Added
            </span>
        </li>
        <li class="changelog-list-item changelog-version-body">
            <h4 id="JavaScript" tabindex="-1">
                JavaScript
                <a class="header-anchor" href="#JavaScript">#</a>
            </h4>
            <ul>
                <li>
                    <p>Code block</p>
                    <article class="message is-warning">
                        <div class="message-body">
                            <p><em>here be dragons</em></p>
                        </div>
                    </article>
                </li>
            </ul>
        </li>
    </div>
</ul>
"
`;

exports[`all the versions are rendered 1`] = `
"<ul class="changelog-list">
    <li class="changelog-list-item is-version">
        <a href="#2.0.0-2021-07-30">
            <span class="anchor" id="2.0.0-2021-07-30"></span>
            <span class="tag is-primary is-large has-text-weight-bold">
                2.0.0
            </span>
        </a>
        <span class="release-date is-uppercase has-text-weight-bold is-size-5">
            30 July 2021
        </span>
    </li>
    <li class="changelog-list-item is-version">
        <a href="#1.1.0-2021-07-28">
            <span class="anchor" id="1.1.0-2021-07-28"></span>
            <span class="tag is-primary is-large has-text-weight-bold">
                1.1.0
            </span>
        </a>
        <span class="release-date is-uppercase has-text-weight-bold is-size-5">
            28 July 2021
        </span>
    </li>
    <li class="changelog-list-item is-version">
        <a href="#1.0.0-2021-07-27">
            <span class="anchor" id="1.0.0-2021-07-27"></span>
            <span class="tag is-primary is-large has-text-weight-bold">
                1.0.0
            </span>
        </a>
        <span class="release-date is-uppercase has-text-weight-bold is-size-5">
            27 July 2021
        </span>
    </li>
</ul>
"
`;

exports[`body of a category is rendered convert from markdown to HTML 1`] = `
"<ul class="changelog-list">
    <li class="changelog-list-item is-version">
        <a href="#1.0.0-2021-07-27">
            <span class="anchor" id="1.0.0-2021-07-27"></span>
            <span class="tag is-primary is-large has-text-weight-bold">
                1.0.0
            </span>
        </a>
        <span class="release-date is-uppercase has-text-weight-bold is-size-5">
            27 July 2021
        </span>
    </li>
    <div>
        <li class="changelog-list-item">
            <span
                class="tag changelog-list-item-category is-medium has-text-weight-bold is-success"
            >
                Added
            </span>
        </li>
        <li class="changelog-list-item changelog-version-body">
            <h4 id="JavaScript" tabindex="-1">
                JavaScript
                <a class="header-anchor" href="#JavaScript">#</a>
            </h4>
            <ul>
                <li>
                    <p>Item 1</p>
                    <ul>
                        <li>Item 1.1</li>
                        <li>Item 1.2</li>
                    </ul>
                </li>
                <li>
                    <p>Item 2</p>
                </li>
                <li>
                    <p>
                        Inline code
                        <code>code</code>
                    </p>
                </li>
                <li>
                    <p>Code block</p>
                    <pre><code class="language-javascript">const foo = 'bar';
</code></pre>
                </li>
                <li>
                    <p>
                        Link
                        <a href="https://example.com">link</a>
                    </p>
                </li>
                <li>
                    <p>
                        Image
                        <img src="https://example.com/image.png" alt="alt" />
                    </p>
                </li>
                <li>
                    <p>
                        Bold
                        <strong>bold</strong>
                    </p>
                </li>
                <li>
                    <p>Item with a description</p>
                    <p>This is a description and it even has a code block:</p>
                    <pre><code class="language-javascript">const foo = 'bar';
</code></pre>
                </li>
            </ul>
        </li>
    </div>
</ul>
"
`;

exports[`known categories are rendered in a fixed order 1`] = `
"<ul class="changelog-list">
    <li class="changelog-list-item is-version">
        <a href="#1.0.0-beta-001-2021-07-29">
            <span class="anchor" id="1.0.0-beta-001-2021-07-29"></span>
            <span class="tag is-primary is-large has-text-weight-bold">
                1.0.0-beta-001
            </span>
        </a>
        <span class="release-date is-uppercase has-text-weight-bold is-size-5">
            29 July 2021
        </span>
    </li>
    <div>
        <li class="changelog-list-item">
            <span
                class="tag changelog-list-item-category is-medium has-text-weight-bold is-success"
            >
                Added
            </span>
        </li>
        <li class="changelog-list-item changelog-version-body">
            <ul>
                <li>Added - Item 1</li>
            </ul>
        </li>
    </div>
    <div>
        <li class="changelog-list-item">
            <span
                class="tag changelog-list-item-category is-medium has-text-weight-bold is-info"
            >
                Changed
            </span>
        </li>
        <li class="changelog-list-item changelog-version-body">
            <ul>
                <li>Changed - Item 1</li>
            </ul>
        </li>
    </div>
    <div>
        <li class="changelog-list-item">
            <span
                class="tag changelog-list-item-category is-medium has-text-weight-bold is-warning"
            >
                Deprecated
            </span>
        </li>
        <li class="changelog-list-item changelog-version-body">
            <ul>
                <li>Deprecated - Item 1</li>
            </ul>
        </li>
    </div>
    <div>
        <li class="changelog-list-item">
            <span
                class="tag changelog-list-item-category is-medium has-text-weight-bold is-danger"
            >
                Removed
            </span>
        </li>
        <li class="changelog-list-item changelog-version-body">
            <ul>
                <li>Removed - Item 1</li>
            </ul>
        </li>
    </div>
    <div>
        <li class="changelog-list-item">
            <span
                class="tag changelog-list-item-category is-medium has-text-weight-bold is-info"
            >
                Fixed
            </span>
        </li>
        <li class="changelog-list-item changelog-version-body">
            <ul>
                <li>Fixed - Item 1</li>
            </ul>
        </li>
    </div>
    <div>
        <li class="changelog-list-item">
            <span
                class="tag changelog-list-item-category is-medium has-text-weight-bold is-info"
            >
                Security
            </span>
        </li>
        <li class="changelog-list-item changelog-version-body">
            <ul>
                <li>Security - Item 1</li>
            </ul>
        </li>
    </div>
</ul>
"
`;

exports[`unknown categories are rendered after known categories 1`] = `
"<ul class="changelog-list">
    <li class="changelog-list-item is-version">
        <a href="#1.0.0-beta-001-2021-07-29">
            <span class="anchor" id="1.0.0-beta-001-2021-07-29"></span>
            <span class="tag is-primary is-large has-text-weight-bold">
                1.0.0-beta-001
            </span>
        </a>
        <span class="release-date is-uppercase has-text-weight-bold is-size-5">
            29 July 2021
        </span>
    </li>
    <div>
        <li class="changelog-list-item">
            <span
                class="tag changelog-list-item-category is-medium has-text-weight-bold is-success"
            >
                Added
            </span>
        </li>
        <li class="changelog-list-item changelog-version-body">
            <ul>
                <li>Item 1</li>
            </ul>
        </li>
    </div>
    <div>
        <li class="changelog-list-item">
            <span
                class="tag changelog-list-item-category is-medium has-text-weight-bold is-info"
            >
                Category A
            </span>
        </li>
        <li class="changelog-list-item changelog-version-body">
            <ul>
                <li>Item 1</li>
            </ul>
        </li>
    </div>
</ul>
"
`;

exports[`unknown categories are rendered in alphabetical order 1`] = `
"<ul class="changelog-list">
    <li class="changelog-list-item is-version">
        <a href="#1.0.0-beta-001-2021-07-29">
            <span class="anchor" id="1.0.0-beta-001-2021-07-29"></span>
            <span class="tag is-primary is-large has-text-weight-bold">
                1.0.0-beta-001
            </span>
        </a>
        <span class="release-date is-uppercase has-text-weight-bold is-size-5">
            29 July 2021
        </span>
    </li>
    <div>
        <li class="changelog-list-item">
            <span
                class="tag changelog-list-item-category is-medium has-text-weight-bold is-info"
            >
                B
            </span>
        </li>
        <li class="changelog-list-item changelog-version-body">
            <ul>
                <li>Item 1</li>
            </ul>
        </li>
    </div>
    <div>
        <li class="changelog-list-item">
            <span
                class="tag changelog-list-item-category is-medium has-text-weight-bold is-info"
            >
                A
            </span>
        </li>
        <li class="changelog-list-item changelog-version-body">
            <ul>
                <li>Item 1</li>
            </ul>
        </li>
    </div>
    <div>
        <li class="changelog-list-item">
            <span
                class="tag changelog-list-item-category is-medium has-text-weight-bold is-info"
            >
                Category C
            </span>
        </li>
        <li class="changelog-list-item changelog-version-body">
            <ul>
                <li>Item 1</li>
            </ul>
        </li>
    </div>
</ul>
"
`;
